src/main/java/io/wispforest/owo/network/OwoHandshake.java

```diff
@@ -96,7 +159,32 @@ private static void syncServer(MinecraftServer server, ServerLoginNetworkHandler
         if (!isAllGood) {
             handler.disconnect(TextOps.concat(PREFIX, Text.of(disconnectMessage.toString())));
         } else {
-            Owo.LOGGER.info("[Handshake] Handshake completed successfully");
+            Owo.LOGGER.info("[LoginHandshake] Handshake completed successfully");
+        }
+    }
+
+    private static void syncServerPlay(MinecraftServer server, ServerPlayerEntity player, ServerPlayNetworkHandler handler, PacketByteBuf buf, PacketSender packetSender) {
+        Owo.LOGGER.info("[PlayHandshake] Receiving client channels");
+
+        IN_PROGRESS_CONNECTIONS.remove(handler.getConnection());
+
+        final var clientChannels = RESPONSE_SERIALIZER.deserializer().apply(buf);
+        final var clientParticleControllers = RESPONSE_SERIALIZER.deserializer().apply(buf);
+
+        StringBuilder disconnectMessage = new StringBuilder();
+
+        boolean isAllGood = verifyReceivedHashes("channels", clientChannels, OwoNetChannel.REGISTERED_CHANNELS, OwoHandshake::hashChannel, disconnectMessage);
+        isAllGood &= verifyReceivedHashes("controllers", clientParticleControllers, ParticleSystemController.REGISTERED_CONTROLLERS, OwoHandshake::hashController, disconnectMessage);
+
+        if (!isAllGood) {
+            // For some reason, this doesn't work if you do it normally
+            var text = TextOps.concat(PREFIX, Text.of(disconnectMessage.toString()));
+            handler.sendPacket(new DisconnectS2CPacket(text));
```

Comment:
Have you tested whether manually sending this packet is actually required? The problem might simply be caused by `disconnect` being called async from a Netty thread if you dont post it

Reply:
I tried doing this with `server.execute`, and it didn't work